crates/core/src/host/host_controller.rs

```diff
@@ -170,27 +168,88 @@ impl HostController {
     /// system, the applications will expect to be called when they are scheduled to be called regardless
     /// of whether the OS has been restarted.
     pub async fn spawn_module_host(&self, mhc: ModuleHostContext) -> Result<ModuleHost, anyhow::Error> {
+        self.setup_module_host(mhc, |mh| async { Ok(mh) }).await
+    }
+
+    /// Set up the [`ModuleHost`] described by the provided [`ModuleHostContext`],
+    /// and run the closure `f` over it.
+    ///
+    /// If `F` returns an `Ok` result, the module host is registered with this
+    /// controller (i.e. [`Self::has_module_host`] returns true), and its
+    /// reducer scheduler is started.
+    ///
+    /// Otherwise, if `F` returns an `Err` result, the module is discarded.
+    async fn setup_module_host<F, Fut, T>(&self, mhc: ModuleHostContext, f: F) -> anyhow::Result<T>
+    where
+        F: FnOnce(ModuleHost) -> Fut,
+        Fut: Future<Output = anyhow::Result<T>>,
+    {
         let key = mhc.dbic.database_instance_id;
 
+        let program_hash = hash_bytes(&mhc.program_bytes);
         let mcc = ModuleCreationContext {
-            dbic: mhc.dbic,
+            dbic: mhc.dbic.clone(),
             scheduler: mhc.scheduler,
-            program_hash: hash_bytes(&mhc.program_bytes),
+            program_hash,
             program_bytes: mhc.program_bytes,
             energy_monitor: self.energy_monitor.clone(),
         };
         let module_host = spawn_rayon(move || Self::make_module_host(mhc.host_type, mcc)).await?;
+        module_host.start();
+
+        let res = f(module_host.clone())
```

Comment:
That's the same, I suppose with `f`. If `f` makes actual changes to the database even if it returns an error, then throwing away the module may still have modified the database. Perhaps we should just add a comment in the description that if and only if `f` returns an error it is guaranteed to have made no mutations to the database.

Reply:
We happen to control `f`, which is carefully crafted to run in a single transaction. But yes, adding more commentary can't hurt.